File: Scala-effekt/Effekt/src/main/scala/Pipes.scala
import effekt._

import scala.util.Random

trait Checker extends Eff {
  def check(array: List[Int]): Op[Unit]
}

object Pipes {

  def check(array: List[Int])(implicit u: Use[Checker]): Control[Unit] = use(u) { u.handler.check(array) }

  def negate(n: Int): Int = {
    if (n == 0) -1 else n
  }

  def findSubArray(array: List[Int])(implicit handler: Use[Checker]): Control[List[Int]] = {
    array match {
      case Nil => pure(Nil)
      case (hd :: tl) => for {
        _ <- check(hd::tl)
        res <- findSubArray(tl)
      } yield res
    }
  }

  def subSearch(sum: Int, acc: List[Int], best: List[Int], array: List[Int]): List[Int] = {
    array match {
      case Nil => best
      case (next :: rest) => {
        val newSum = sum + negate(next)
        val newAcc = acc ::: List(next)
        if (newSum == 0) subSearch(newSum, newAcc, newAcc, rest)
        else subSearch(newSum, newAcc, best, rest)
      }
    }
  }

  def pipesHandler = new Handler.Basic[List[Int], List[Int]] with Checker {
    override def check(array: List[Int]): Unit => (Unit => Unit => Control[List[Int]]) => Control[List[Int]] =
      _ => resume => {
        val currentBest = subSearch(0, Nil, Nil, array)
        for {
          nextBest <- resume()()
        } yield if (currentBest.length > nextBest.length) currentBest else nextBest
      }

    override def unit: List[Int] => List[Int] = identity
  }

  def generate(i: Int): List[Int] = {
    val r = Random
    def gen(n: Int, acc: List[Int]): List[Int] = {
      if (n == 0) acc
      else gen(n - 1, r.nextInt(2)::acc)
    }
    gen(i, Nil)
  }

  def run(problem: List[Int]): List[Int] = pipesHandler { implicit h => findSubArray(problem) }.run()

  def main(args: Array[String]): Unit = {
    if (args.length >= 1) {
      val n = args(0).toInt
      val t = generate(n)
      val now = System.currentTimeMillis()
      run(t)
      val time = System.currentTimeMillis() - now
      println(time)
    }
  }
}

File: OCaml-multicore/Pipes/pipes.ml
module Pipes : sig
  val solve : int list -> int list
end =
struct
  effect Check : int list -> unit

  let negate = function | 0 -> -1 | n -> n

  let rec find_subarray = function
  | []    -> []
  | x::xs -> perform (Check (x::xs)); find_subarray xs

  let solve (lst : int list) =
    try find_subarray lst with
    | effect (Check sublst) k ->
      let rec sub_search sum acc best = function
        | []         -> best
        | next::rest ->
          let newsum = sum + (negate next) in
          let newacc  = acc @ [next] in
          if newsum == 0 then
            sub_search newsum newacc newacc rest
          else
            sub_search newsum newacc best rest
      in
      let current_best = sub_search 0 [] [] sublst in
      let next_best = continue k () in
      if List.length current_best > List.length next_best then
        current_best
      else
        next_best
end
